src/storage/exec/IndexScanNode.cpp

```diff
@@ -0,0 +1,578 @@
+  std::string val;
+  bool greater = !includeStart_;
+  CHECK_NE(value.type(), Value::Type::NULLVALUE);
+  if (colDef.get_type() == ::nebula::meta::cpp2::PropertyType::GEOGRAPHY) {
+    val = value.getStr();
+  } else if (value.type() == Value::Type::STRING) {
+    bool truncated = false;
+    val = encodeString(value, *colDef.get_type_length(), truncated);
+    greater &= !truncated;
+    if (UNLIKELY(truncated)) {
+      QFList_.emplace_back([val, startPos = offset](const std::string& k) {
+        int ret = memcmp(val.data(), k.data() + startPos, val.size());
+        DVLOG(1) << folly::hexDump(val.data(), val.size());
+        DVLOG(1) << folly::hexDump(k.data(), k.size());
+        CHECK_LE(ret, 0);
+        return ret == 0 ? Qualified::UNCERTAIN : Qualified::COMPATIBLE;
+      });
+    }
+  } else if (value.type() == Value::Type::FLOAT) {
+    bool isNaN = false;
+    val = encodeFloat(value, isNaN);
+    greater |= isNaN;
+    // TODO(hs.zhang): Optimize the logic of judging NaN
+    QFList_.emplace_back([startPos = offset, length = val.size()](const std::string& k) {
+      int ret = memcmp("\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF", k.data() + startPos, length);
+      return ret == 0 ? Qualified::INCOMPATIBLE : Qualified::COMPATIBLE;
+    });
+  } else {
+    val = IndexKeyUtils::encodeValue(value);
+  }
+  if (greater) {
+    val.append(suffixLength_ + 1, '\xFF');
+  }
+  key += val;
+  return val;
+}
+std::string RangePath::encodeEndValue(const Value& value,
+                                      const ColumnTypeDef& colDef,
+                                      std::string& key,
+                                      size_t offset) {
+  CHECK_NE(value.type(), Value::Type::NULLVALUE);
+  std::string val;
+  bool greater = includeEnd_;
+  if (colDef.get_type() == ::nebula::meta::cpp2::PropertyType::GEOGRAPHY) {
+    val = value.getStr();
+  } else if (value.type() == Value::Type::STRING) {
+    bool truncated = false;
+    val = encodeString(value, *colDef.get_type_length(), truncated);
+    greater |= truncated;
+    if (UNLIKELY(truncated)) {
+      QFList_.emplace_back([val, startPos = offset](const std::string& k) {
+        int ret = memcmp(val.data(), k.data() + startPos, val.size());
+        DVLOG(3) << folly::hexDump(val.data(), val.size());
+        DVLOG(3) << folly::hexDump(k.data(), k.size());
+        CHECK_GE(ret, 0);
+        return ret == 0 ? Qualified::UNCERTAIN : Qualified::COMPATIBLE;
+      });
+    }
+  } else if (value.type() == Value::Type::FLOAT) {
+    bool isNaN = false;
+    val = encodeFloat(value, isNaN);
+    greater |= isNaN;
+    if (UNLIKELY(isNaN)) {
+      QFList_.emplace_back([startPos = offset, length = val.size()](const std::string& k) {
+        DVLOG(3) << "check NaN:" << startPos << "\t" << length;
+        DVLOG(3) << '\n' << folly::hexDump(k.data(), k.size());
+        int ret = memcmp("\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF", k.data() + startPos, length);
```

Comment:
IIRC . It seems that the Graph layer has duplicate code.

Reply:
Yes, there is similar logic in graph. But that part of the logic in the graph is wrong, which is why this PR is mentioned. The processing logic must be moved to storage to ensure the correctness of the query results. I will add some notes later to explain the relevant content